Improve registries documentation (#429)
microsoft/vcpkg-docs: vcpkg/maintainers/registries.md

| -------------- | -------------- | ----------- |
| `default`      | BaselineObject | The default baseline, required for Git registries. |
| Named baseline | BaselineObject | Additional baselines. The field name corresponds to the baseline name. |

*BaselineObject*

The baseline object is a dictionary, with each key corresponding to a port name
in the registry and its value being the latest version of the port.

| Name      | Type                  | Description |
| --------- | --------------------- | ----------- |
| Port name | BaselineVersionObject | A mapping of a port name to its latest version |

*BaselineVersionObject*

| Name | Type | Description |
| ---- | ---- | ----------- |
| `baseline` | string | A string corresponding to the latest available version of the port in the registry. |
| `port-version` | integer | An integer corresponding to the latest port version of the port in the registry |

#### Example of a `baseline.json` file in a Git registry

In a registry containing a single port named `foo` at version `1.0.0#1`, the
`baseline.json` file contents should be:

```json
{
  "default": {
    "foo": { 
      "baseline": "1.0.0", 

>       "port-version": "1"

Reviewer: ```suggestion
      "port-version": 1
```

port-version is a (in the JSON sense) number, not a string

    }
  }
}
```

### <a name="git-version-file"></a> Version file layout in Git registries

The `versions` directory contains all the information about which versions of
packages are contained in the registry, along with the method to retrieve those
versions from the repository's history.

*Top-level fields*

| Name       | Type            | Description |
| ---------- | --------------- | --------------------------- |
| `versions` | VersionObject[] | An array of version objects. Contains an entry for each version of the port in the history of the registry. |

*VersionObject*

| Name       | Type   | Description |
| ---------- | ------ | ----------- |

> | `git-tree` | string | A Git calculated SHA used to retrieve the port version |

Reviewer: ```suggestion
| `git-tree` | string | A git tree SHA used to retrieve the port contents |
```

Since we get more than the version out of here I think it's good to describe that we're getting the entire port contents from here.

| [`version`<br>`version-semver`<br>`version-date`<br>`version-string`](../reference/vcpkg-json.md#version) | string | Upstream version information |
| [port-version](../reference/vcpkg-json.md#port-version) | integer | Port files revision |

> 

Reviewer: I think we should add something like:

> The `version` / `verison-semver` / `version-date` / `version-string` and `port-version` match the values in the port retrieved by the `git-tree`. For example:
> ```console
> PS D:\vcpkg> Get-Content D:\vcpkg\versions\c-\curl.json -TotalCount 7
> {
>   "versions": [
>     {
>       "git-tree": "6ef1763f3cbe570d6378632c9b5793479c37fb07",
>       "version": "8.11.0",
>       "port-version": 1
>     },
> PS D:\vcpkg> git show 6ef1763f3cbe570d6378632c9b5793479c37fb07:vcpkg.json | Select-Object -First 5
> {
>   "name": "curl",
>   "version": "8.11.0",
>   "port-version": 1,
>   "description": "A library for transferring data with URLs",
> ```

I'm torn on whether this is a 'must' situation. It's true that if a user is manually doing this they must ensure that they do this, but it's also true that we provide tooling that makes this an implementation detail most of the time.

Author: Maybe we can add this to the "Recommended practices" section in the conceptual article.

> #### Example of a Git registry version file

Reviewer 2: ```suggestion
#### Example of a Git registry versions file
```

```json
{
  "versions": [
    {
      "git-tree": "963060040c3ca463d17136e39c7317efb15eb6a5",
      "version": "1.2.0",
      "port-version": 0
    },
    {
      "git-tree": "548c90710d59c174aa9ab10a24deb69f1d75ff8f",
      "version": "1.1.0",
      "port-version": 0
    },
    {
      "git-tree": "67d60699c271b7716279fdea5a5c6543929eb90e",
      "version": "1.0.0",
      "port-version": 0
    }
  ]
}
``` 

#### <a name="obtain-git-tree-sha"></a> Obtaining a `git-tree` SHA

vcpkg uses Git's capabilities to retrieve specific versions of a port contained
in its commit history. The method used is to retrieve the `git-tree` object from

> the repository as specified in the version file of a port.

Reviewer 2: ```suggestion
the repository as specified in the versions file of a port.
```

Each port directory in a Git registry has a unique SHA associated with it

> (referred as `git-tree` in the version files). The SHA is calculated using the

Reviewer 2: ```suggestion
(referred as `git-tree` in the versions files). The SHA is calculated using the
```

ports in the registry to their baseline version.

| Name           | Type           | Description |
| -------------- | -------------- | ----------- |
| Named baseline | BaselineObject | Additional baselines. The field name corresponds to the baseline name. |

*BaselineObject*

The baseline object is a dictionary, with each key corresponding to a port name
in the registry and its value being the latest version of the port.

| Name      | Type                  | Description |
| --------- | --------------------- | ----------- |
| Port name | BaselineVersionObject | A mapping of a port name to its latest version |

*BaselineVersionObject*

| Name | Type | Description |
| ---- | ---- | ----------- |
| `baseline` | string | A string corresponding to the latest available version of the port in the registry. |
| `port-version` | integer | An integer corresponding to the latest port version of the port in the registry |

The layout of the baseline file in a filesystem registry is the same as for [Git
registries](#git-baseline). The only difference being that filesystems
don't require a `default` baseline.

#### Example of a `baseline.json` file in a Git registry

```json
{

>   "my-baseline": {

Reviewer: I think this was better before when `my-baseline` was clearly a baseline revision rather than `my-baseline`.

    "foo": {
      "baseline": "1.0.0",
      "port-version": 1
    }
  }
}
```
### <a name="filesystem-version-file"></a> Version file layout in Git registries

The `versions` directory contains all the information about which versions of
packages are contained in the registry, along with the method to retrieve those
versions from a filesystem location.

*Top-level fields*

| Name       | Type            | Description |
| ---------- | --------------- | --------------------------- |
| `versions` | VersionObject[] | An array of version objects. Contains an entry for each version of the port in the registry. |

*VersionObject*

| Name       | Type   | Description |
| ---------- | ------ | ----------- |
| `path`     | string | The filesystem location where the port files for the corresponding version are located |
| [`version`<br>`version-semver`<br>`version-date`<br>`version-string`](../reference/vcpkg-json.md#version) | string | Upstream version information |
| [port-version](../reference/vcpkg-json.md#port-version) | integer | Port files revision |

> 

Reviewer: Ditto showing an example that these need to match might be nice.

When specifying the `path` of a registry, the `$` character can be used to
reference the root of the registry. Otherwise, absolute paths can be used
instead.

#### Example of a filesystem registry version file

```json
{
  "versions": [
    {
      "git-tree": "$/ports/foo/1.2.0",
      "version": "1.2.0",
      "port-version": 0
    },
    {

>       "git-tree": "$/ports/foo/1.1.0",

Reviewer: This should be `"path"` https://github.com/microsoft/vcpkg-tool/blob/b8680ed15eb1350e420a18ce37556d3dc9e6d084/src/vcpkg/registries.cpp#L117
